File: Data Structures & Algorithms/merge-k-sorted-linked-lists/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists or len(lists) < 1:
            return None
        l1 = None
        for i in range(0, len(lists)):
            l2 = lists[i]
            # print(l1, l2)
            l1 = self.mergeSortedList(l1, l2)
            # print(mergedList)
        return l1

    def mergeSortedList(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        result = curr = ListNode()
        # print(1)
        while l1 and l2:
            if l1.val <= l2.val:
                curr.next = l1
                l1 = l1.next
            else:
                curr.next = l2
                l2 = l2.next
            curr = curr.next

        while l1:
            curr.next = l1
            l1 = l1.next
            curr = curr.next

        while l2:
            curr.next = l2
            l2 = l2.next
            curr = curr.next

        # curr.next = None

        # print(result)
        return result.next



        

File: Data Structures & Algorithms/add-two-numbers/submission-1.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */

class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode r1 = new ListNode(-1);
        ListNode head = r1;
        int carry = 0;
        while (l1 != null && l2 != null) {
            int sum = carry + l1.val + l2.val;
            carry = sum / 10;
            int digit = sum % 10;
            ListNode node = new ListNode(digit);
            r1.next = node;
            r1 = node;
            l1 = l1.next;
            l2 = l2.next;
        }

        while (l1 != null) {
            int sum = carry + l1.val;
            carry = sum / 10;
            int digit = sum % 10;
            ListNode node = new ListNode(digit);
            r1.next = node;
            r1 = node;
            l1 = l1.next;
        }

        while (l2 != null) {
            int sum = carry + l2.val;
            carry = sum / 10;
            int digit = sum % 10;
            ListNode node = new ListNode(digit);
            r1.next = node;
            r1 = node;
            l2 = l2.next;
        }

        if (carry > 0) {
            ListNode node = new ListNode(carry);
            r1.next = node;
        }

        return head.next;
    }
}
